### DISCUSSION TITLE
MESA Adaptive Moving Average (MAMA)

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#211] [Indicators]
Community discussion about [MESA Adaptive Moving Average (MAMA)](https://dotnet.stockindicators.dev/indicators/Mama/#content)

**Comment by DaveSkender** [comment]
I think John Ehlers is just making up words [in his formula](http://mesasoftware.com/papers/MAMA.pdf) for this indicator.  Remember, you'll need to determine the quadrature before the homodyne discriminator! 😮 Serious math in this one ... he is definitely a genius to be able to create this one.  I now understand why he nicknamed it "mother of adaptive moving averages".

**Comment by DaveSkender** [comment]
Does anyone know of any [free] professionally generated MESA charting websites?  I've developed this indicator, done the manual calculations based on Ehler's paper, but can't seem to find other reputable sources for the final [comparison] validation that I like to do.  This indicator will be available in version 1.4.0.

As a side note, if anyone wants to review the code and provide feedback, please do.  It's not yet released in the public library, but it is available in the [internal package manager](https://dev.azure.com/skender/Stock.Indicators/_packaging?_a=package&feed=Internal&package=Skender.Stock.Indicators&protocolType=NuGet) for anyone who wants to beta test it.

**Reply by Fredrik-C** [reply]
Perhaps you can benchmark against Alpha Vantage?

https://www.alphavantage.co/documentation/#mama

While you are at it, please implement _Hilbert transform instantaneous trendline_ :)

**Reply by DaveSkender** [reply]
Added the transform to the backlog #350 
I’ll checkout AlphaVantage API for comparison, thx.

**Reply by DaveSkender** [reply]
For anyone following this, I've done this comparison to the Alpha Vantage API for MESA and have successfully used it as a final validation of the algorithm.  This was in addition to the original validation based on manual calculations.

One thing to note about the Alpha Vantage API is that they have chosen to use `Close` price instead of the original Ehler's formula `(High+Low)/2` price.  The difference between using the two alternative prices is small.  This library uses and is true to the [original formula](http://mesasoftware.com/papers/MAMA.pdf).

**Comment by JosueNina** [comment]
Hi Dave,

I'm currently working on creating my own [MAMA](https://www.mesasoftware.com/papers/MAMA.pdf) indicator and testing it with different libraries. While comparing your indicator with the MAMA from the ta-lib library, I noticed that the values are slightly different. Could you explain if this is due to using a simple average for the first 6 values? I found the following code in your [implementation](https://github.com/DaveSkender/Stock.Indicators/blob/main/src/m-r/Mama/Mama.Series.cs):

```
sumPr += pr[i];

if (i == 5)
{
    r.Mama = (sumPr / 6d).NaN2Null();
    r.Fama = r.Mama;
}
```
Additionally, I noticed that you use radians for the period calculation, which I think is correct:
```
pd[i] = im[i] != 0 && re[i] != 0
                    ? 2 * Math.PI / Math.Atan(im[i] / re[i])
                    : 0d;
```
And then you use degrees for the phase calculation:
`ph[i] = (i1[i] != 0) ? Math.Atan(q1[i] / i1[i]) * 180 / Math.PI : 0;`

Could you explain why this distinction is made?

Thanks!

**Reply by DaveSkender** [reply]
I’m going to have to take a minute to think through this.  Also, I’ll not be able to really explain what creative things `ta-lib` had done, but can confirm a few of these things against the [original author’s publication](https://mesasoftware.com/papers/MAMA.pdf).  I suspect this is more of a math problem than anything else.  Tracking in #1278

**Reply by DaveSkender** [reply]
~My gut says that `i1[i]` is in degree units and `180 / Math.PI` is converting it back to radians, and I should have put my parenthesis in a different location.~

[bad guess, will need to take a deeper look at formula and [manual calcs](https://github.com/DaveSkender/Stock.Indicators/blob/f8124dc1715e0f515823f9d6fa040051b1e96243/tests/indicators/m-r/Mama/Mama.Calc.xlsx)]

**Reply by DaveSkender** [reply]
Regarding `SMA` initialization.  Since this indicator uses a smoothing technique, `FAMA` and `MAMA` values have to be initialized.  John Ehler's whitepaper implies that these initialize as `0` values.  I usually do honor the original author's approach; however, in this case, using `SMA` is more aligned with how most moving averages are initialized to reduce early period errors.

<img width="325" alt="image" src="https://github.com/user-attachments/assets/08a410a9-55d4-4890-879f-d20d178e1266">

---

Using the `SMA` initialization value here instead of `0` allows the early periods to have a smaller convergence errors, even though they do end up converging synchronously from the 100-350+ periods.[^1]

<img width="350" alt="image" src="https://github.com/user-attachments/assets/0aa23ae1-47d5-4281-9fb3-c7a34cd531c9">

---

In reality, this is what initial values look like, comparing `SMA` (gray line) vs `0` (orange line) initialization values to steady-state converged `MAMA` values (blue line).

![](https://github.com/user-attachments/assets/b005fba7-bbd8-4107-9ab9-4f1809a804ac)

[^1]: This convergence analysis uses different quotes, price scale (potentially confusing).

**Reply by DaveSkender** [reply]
I'm still looking into the radian vs. degrees concern; though still believe this is just a way to get everything into the correct units for each value used in the final calculations.

**Reply by JosueNina** [reply]
That makes sense. As for the second part, I've seen many implementations doing `360 / arctan()`, which is incorrect. In your case, `2 * pi / arctan() `makes sense since both are in radians. Also, the phase should be in degrees
![image](https://github.com/user-attachments/assets/ed519eaf-6e1c-4230-ac44-f48f2805cffd)
, which your code also handles correctly.

I think that answers all my questions. Thanks again for the quick response!

**Reply by DaveSkender** [reply]
Mostly depends on the code language.  .NET returns radians from `math.Atan()` while others may be in degrees.
